[Messenger] Add `ListableReceiverInterface` support to `RedisReceiver`

mudassaralichouhan opened the pull request:
| Q             | A
|---------------|-----
| Branch?       | 8.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #61436
| License       | MIT
| Doc PR        | symfony/symfony-docs#...

### Description

This PR adds support for the `ListableReceiverInterface` to the Redis Messenger transport. It allows listing and finding messages in Redis-backed queues, enabling integrations with tools like `zenstruck/messenger-monitor-bundle`.

### Features

- `RedisReceiver` now implements `ListableReceiverInterface`
- Uses Redis stream XRANGE commands
- New methods in `Connection`: `findAll()`, `find($id)`
- Fully tested (unit + functional)
- Backward-compatible and non-breaking

### Usage

```php
$receiver->all();          // Returns all pending messages
$receiver->find('123-0');  // Returns a specific message by Redis stream ID

mudassaralichouhan committed:
[Messenger] Add `ListableReceiverInterface` support to `RedisReceiver`

fabpot commented on src/Symfony/Component/Messenger/Bridge/Redis/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * Implement the `CloseableTransportInterface` to allow closing the Redis connection
  * Implement the `KeepaliveReceiverInterface` to enable asynchronously notifying Redis that the job is still being processed, in order to avoid timeouts
  * Add support for `valkey:` / `valkeys:` schemes
+ * Implement the `ListableReceiverInterface` to enable listing and finding messages in Redis streams
```
Should be moved to a new 7.4 section.

nicolas-grekas replied:
should be moved to a new 7.4 section

fabpot commented on src/Symfony/Component/Messenger/Bridge/Redis/README.md:
````diff
@@ -3,6 +3,39 @@ Redis Messenger
+// Find a specific message by ID
+$message = $receiver->find('1234567890-0');
+```
+
````
```suggestion
```

This should be part of the docs under symfony/symfony-docs instead.

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/RedisReceiver.php:
```diff
@@ -105,6 +106,83 @@ public function getMessageCount(): int
+        }
+
+        try {
+            if (\array_key_exists('body', $redisEnvelope) && \array_key_exists('headers', $redisEnvelope)) {
```
there's some duplication now
can't we do like the Doctrinereceiver, add a private createEnvelopeFromData method?

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -736,6 +736,85 @@ private function getRedis(): \Redis|Relay|\RedisCluster
+            throw new TransportException($e->getMessage(), 0, $e);
+        }
+
+        if (false === $range || empty($range)) {
```
```suggestion
        if (!$range) {
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -736,6 +736,85 @@ private function getRedis(): \Redis|Relay|\RedisCluster
+        }
+
+        if (false === $range || empty($range)) {
+            if ($error = $redis->getLastError() ?: null) {
```
```suggestion
            if ($error = $redis->getLastError()) {
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -736,6 +736,85 @@ private function getRedis(): \Redis|Relay|\RedisCluster
+            return null;
+        }
+
+        $data = reset($range);
```
```suggestion
        $data = current($range);
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -736,6 +736,85 @@ private function getRedis(): \Redis|Relay|\RedisCluster
+        }
+
+        if (false === $range) {
+            if ($error = $redis->getLastError() ?: null) {
```
```suggestion
            if ($error = $redis->getLastError()) {
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/RedisReceiver.php:
```diff
@@ -105,6 +106,83 @@ public function getMessageCount(): int
+                } else {
+                    $envelope = $this->serializer->decode($redisEnvelope);
+                }
+            } catch (MessageDecodingFailedException $exception) {
```
```suggestion
            } catch (MessageDecodingFailedException) {
```

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/RedisReceiver.php:
```diff
@@ -105,6 +106,83 @@ public function getMessageCount(): int
+            } else {
+                $envelope = $this->serializer->decode($redisEnvelope);
+            }
+        } catch (MessageDecodingFailedException $exception) {
```
```suggestion
        } catch (MessageDecodingFailedException) {
```

nicolas-grekas reviewed (commented):
Looks nice, I just have CS-related comments.

mudassaralichouhan commented:
Hi @fabpot, all requested changes have been addressed. Could you please re-review when you have a chance?

fabpot commented on src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ListableReceiverFunctionalTest.php:
```diff
@@ -0,0 +1,73 @@
+namespace Symfony\Component\Messenger\Bridge\Redis\Tests\Transport;
+
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Messenger\Bridge\Redis\Tests\Fixtures\DummyMessage;
```
```suggestion
```

fabpot commented on src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ListableReceiverFunctionalTest.php:
```diff
@@ -0,0 +1,73 @@
+use Symfony\Component\Messenger\Bridge\Redis\Transport\RedisReceiver;
+
+use Symfony\Component\Messenger\Transport\Serialization\PhpSerializer;
+
```
```suggestion
```

fabpot commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -172,7 +179,12 @@ public function __construct(array $options, \Redis|Relay|\RedisCluster|null $red
+     *
+     * @return \Redis
+     */
+    private static function initializeRedis($redis, string $host, int $port, string|array|null $auth, array $params)
```
Can you remove the phpdocs and type hint the function signature directly instead?

fabpot commented:
Tests seem broken by your changes.
You must also rebase your PR to get rid of the merge commit.

mudassaralichouhan commented:
> Tests seem broken by your changes. You must also rebase your PR to get rid of the merge commit.

You can now comment on the PR that the rebase is done and tests pass.

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -634,8 +672,12 @@ public function keepalive(string $id, ?int $seconds = null): void
             );
-        } catch (\RedisException|\Relay\Exception $e) {
-            throw new TransportException($e->getMessage(), 0, $e);
+        } catch (\Throwable $e) {
```
why these changes? They should all be reverted to me

ro0NL commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -54,7 +54,8 @@ class Connection
     ];
 
-    private \Redis|Relay|\RedisCluster|null $redis = null;
+    /** @var \Redis|\RedisCluster|null */
```
> Backward-compatible and non-breaking

😕

ro0NL commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -713,8 +755,12 @@ private function rawCommand(string $command, ...$arguments): mixed
             } else {
                 $result = $redis->rawCommand($command, $this->queue, ...$arguments);
             }
-        } catch (\RedisException|\Relay\Exception $e) {
-            throw new TransportException($e->getMessage(), 0, $e);
+        } catch (\Throwable $e) {
```
why the exception swap?

nicolas-grekas replied:
@mudassaralichouhan you might want to review your PR here on github:
many changes don't make much sense - or have to be justified - Relay-related

ro0NL commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -736,6 +736,78 @@ private function getRedis(): \Redis|Relay|\RedisCluster
+    {
+        if ($this->autoSetup) {
+            $this->setup();
+        }
```
i dont think this is needed

ro0NL commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -452,7 +452,7 @@ public function get(): ?array
 
         if (false === $messages) {
-            if ($error = $redis->getLastError() ?: null) {
+            if ($error = $redis->getLastError()) {
```
related?

ro0NL reviewed (commented):
much easier to review now ;)

fabpot commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -736,6 +736,74 @@ private function getRedis(): \Redis|Relay|\RedisCluster
+            throw new TransportException($e->getMessage(), 0, $e);
+        }
+
+        if (false === $range || empty($range)) {
```
```suggestion
        if (!$range) {
```

We try to avoid using `empty` in the Symfony codebase.

fabpot commented:
@mudassaralichouhan Can you rebase on current 7.4 (first, to remove the merge commit, then to see if the tests pass). Thank you.

nicolas-grekas commented on src/Symfony/Component/Messenger/Bridge/Redis/Transport/Connection.php:
```diff
@@ -751,6 +751,74 @@ private function getRedis(): \Redis|Relay|\RedisCluster
         return $this->redis;
     }
 
+    public function findAll(?int $limit = null): array
```
I can't find any usages of the findAll method, even for the Doctrine bridge.
I guess it can be removed? /cc @kbond

kbond replied:
Not sure what you mean? With the Doctrine bridge it's used in `DoctrineReceiver::all()`.

fabpot reviewed (changes requested):
Tests seem broken by the changes from this PR.

mudassaralichouhan commented:
I’ve fixed the Redis `findAll()` implementation and updated the tests — everything is green locally now.

fabpot reviewed (changes requested):
Please, remove all comments that are not useful and just describe what the code does, it's not useful.

Overall, I feel like the tests do not test a lot as everything is mocked.

xabbuh commented on src/Symfony/Component/Messenger/Bridge/Redis/Tests/Transport/ListableReceiverFunctionalTest.php:
```diff
@@ -0,0 +1,70 @@
+    public function testEndToEndListableReceiverFunctionality()
+    {
+        // Create a mock Redis connection
+        $connection = $this->createMock(Connection::class);
```
Mocking the Redis connection does not make much sense to me if the test is an integration test. It should use a real Redis server instead.

mudassaralichouhan commented:
Thanks for the feedback. I’ve removed the old mock-based tests with unnecessary comments and replaced them with `RedisListableIntegrationTest`, which uses a real Redis server. This should cover the functionality in a more meaningful way.

nicolas-grekas reviewed (approved):
I rebase on 8.1, simplified tests and fixed a few CS-related things

nicolas-grekas commented:
Thank you @mudassaralichouhan.
